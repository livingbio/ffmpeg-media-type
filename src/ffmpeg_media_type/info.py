import os
from typing import Literal
from urllib.parse import urlparse

from pydantic import BaseModel

from .utils.ffmpeg import FFProbeInfo, ffprobe, get_ffmpeg, load_cache
from .utils.shell import call


class MediaInfo(BaseModel):
    type: Literal["image", "video", "audio"]

    width: int | None = None
    height: int | None = None
    duration: float | None = None

    format: str | None = None
    size: int | None = None
    suggest_ext: str | None = None


def generate_thumbnail(video_path: str, thumbnail_path: str, time_offset: float = 0) -> str:
    ffmpeg_cmd = get_ffmpeg() + [
        "-y",  # Overwrite output file if it exists
        "-i",
        video_path,  # Input video path
        "-ss",
        str(time_offset),  # Time offset (seek to the specified position)
        "-vframes",
        "1",  # Number of frames to output
        "-vf",
        "scale=320:-1",  # Thumbnail size (width: 320, height: proportional)
        "-q:v",
        "2",  # Quality (2 - high, 5 - low)
        thumbnail_path,  # Output thumbnail path
    ]

    call(ffmpeg_cmd)

    return thumbnail_path


def _extract_file_extension(uri: str) -> str:
    parsed_uri = urlparse(uri)
    path = parsed_uri.path
    filename = os.path.basename(path)
    _, extension = os.path.splitext(filename)
    return extension[1:].lower()


_KNOWN_CODEC_EXTS = {
    "png_pipe": ["png"],
    "svg_pipe": ["svg"],
    "tiff_pipe": ["tiff", "tif"],
    "bmp_pipe": ["bmp"],
    "gif_pipe": ["gif"],
    "jpeg_pipe": ["jpeg", "jpg"],
    "webp_pipe": ["webp"],
    "mjpeg": ["jpg", "jpeg", "mjpeg"],
}


def _guess_media_info(
    uri: str,
    info: FFProbeInfo,
) -> MediaInfo:
    infos = load_cache()
    current_ext = _extract_file_extension(uri)

    format_name = info.format.format_name
    duration = info.format.duration

    # NOTE: handle ffmpeg's image compatibility
    if format_name == "image2":
        format_name = info.streams[0].codec_name

    # NOTE: detect file extension
    if format_name in _KNOWN_CODEC_EXTS:
        common_exts = _KNOWN_CODEC_EXTS[format_name]
    elif format_name in infos:
        common_exts = infos[format_name].common_exts
    else:
        common_exts = []

    if current_ext in common_exts:
        suggest_ext = current_ext
    elif common_exts:
        suggest_ext = common_exts[0]
    else:
        suggest_ext = None

    # NOTE: we classify gif as image
    if not duration or format_name in ("gif", "mjpeg"):
        return MediaInfo(
            type="image",
            width=info.streams[0].width or 0,
            height=info.streams[0].height or 0,
            duration=duration or 0,
            format=format_name,
            size=info.format.size,
            suggest_ext=suggest_ext,
        )

    for stream in info.streams:
        # NOTE: if there is at least one video stream, the media is video
        if stream.codec_type == "video" and format_name not in ("mp3",):
            width = stream.width
            height = stream.height

            return MediaInfo(
                type="video",
                width=width or 0,
                height=height or 0,
                duration=duration or 0,
                format=format_name,
                size=info.format.size,
                suggest_ext=suggest_ext,
            )

    # NOTE: if there is no video stream, the media is audio
    return MediaInfo(
        type="audio",
        width=0,
        height=0,
        duration=duration or 0,
        format=format_name,
        size=info.format.size,
        suggest_ext=suggest_ext,
    )


def detect(uri: str) -> MediaInfo:
    probe_info = ffprobe(uri)
    return _guess_media_info(uri, probe_info)
