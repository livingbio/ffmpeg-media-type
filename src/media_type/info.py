from dataclasses import dataclass
from typing import Literal


@dataclass
class MediaInfo:
    type: Literal["image", "video", "audio"]

    width: int | None = None
    height: int | None = None
    duration: float | None = None

    format: str | None = None
    size: int | None = None
    suggest_ext: str | None = None


def detect(uri: str) -> None:
    ...
