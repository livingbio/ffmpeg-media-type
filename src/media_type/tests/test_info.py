from pathlib import Path

import pytest
from syrupy.assertion import SnapshotAssertion
from syrupy.filters import props

from ..info import detect


@pytest.mark.parametrize(
    "case",
    [pytest.param(k, id=k.name) for k in (Path(__file__).parent.parent / "utils/tests").glob("test_ffmpeg/*")],
)
def test_detect(case: Path, snapshot: SnapshotAssertion) -> None:
    snapshot(name=case.name, exclude=props("duration")) == detect(str(case)).dict()
