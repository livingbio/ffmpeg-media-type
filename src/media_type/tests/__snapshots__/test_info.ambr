# serializer version: 1
# name: test_detect[10257640.mp4]
  MediaInfo(type='video', width=640, height=480, duration=6.803447, format='mov,mp4,m4a,3gp,3g2,mj2', size=593915, suggest_ext='mp4')
# ---
# name: test_detect[188794898-eeb81e4b-b33a-4bdf-85a8-3b4b8460b88a.png]
  MediaInfo(type='image', width=262, height=259, duration=0.0, format='png_pipe', size=47210, suggest_ext='png')
# ---
# name: test_detect[201432913186_fish.gif]
  MediaInfo(type='image', width=480, height=360, duration=1.2, format='gif', size=30333, suggest_ext='gif')
# ---
# name: test_detect[4ox1IBJxK3cVd0uJA6gOtceSP3g.jpg]
  MediaInfo(type='video', width=1254, height=836, duration=0.04, format='mjpeg', size=1249892, suggest_ext='mjpg')
# ---
# name: test_detect[9b5174019e8c9bec190903528ff6d0392a2d6a9b.mp4]
  MediaInfo(type='video', width=1920, height=1080, duration=38.740617, format='mov,mp4,m4a,3gp,3g2,mj2', size=7164941, suggest_ext='mp4')
# ---
# name: test_detect[9c7ee468-24a3-4930-9832-73f0f6001ab1.webp]
  MediaInfo(type='image', width=1920, height=787, duration=0.0, format='webp_pipe', size=55242, suggest_ext='webp')
# ---
# name: test_detect[Cle9pQ5.jpg]
  MediaInfo(type='video', width=720, height=668, duration=0.04, format='mjpeg', size=72973, suggest_ext='mjpg')
# ---
# name: test_detect[SampleAudio_0.4mb.mp3]
  MediaInfo(type='audio', width=0, height=0, duration=27.74475, format='mp3', size=443926, suggest_ext='mp3')
# ---
# name: test_detect[SampleVideo_1280x720_1mb.mp4]
  MediaInfo(type='video', width=1280, height=720, duration=5.312, format='mov,mp4,m4a,3gp,3g2,mj2', size=1055736, suggest_ext='mp4')
# ---
# name: test_detect[cd89b1bf161582d55161f5ba94d03cc7728825c2.jpe]
  MediaInfo(type='image', width=4716, height=3144, duration=0.0, format='jpeg_pipe', size=12642182, suggest_ext='jpeg')
# ---
# name: test_detect[maxresdefault.jpg-sdafdeagwdsadf]
  MediaInfo(type='image', width=1280, height=720, duration=0.0, format='jpeg_pipe', size=99802, suggest_ext='jpeg')
# ---
# name: test_detect[overlay.png]
  MediaInfo(type='image', width=163, height=117, duration=0.0, format='png_pipe', size=2212, suggest_ext='png')
# ---
# name: test_detect[silent_video.mp4]
  MediaInfo(type='video', width=1920, height=1080, duration=6.965292, format='mov,mp4,m4a,3gp,3g2,mj2', size=14770522, suggest_ext='mp4')
# ---
